tests/integration/app-charm/src/charm.py

```diff
@@ -58,6 +59,34 @@ def on_subject_allowed_admin(self, event: SubjectAllowedEvent):
         logger.info(f"{event.username} {event.password} {event.endpoints} {event.tls}")
         return
 
+    def _get_credentials_action(self, event: ActionEvent) -> None:
+        """Handler for get-credentials action."""
+        user = event.params.get("username")
+        username = ""
+        password = ""
+        relation = self.model.get_relation(
+            relation_name=REL_NAME_USER if user == "user" else REL_NAME_ADMIN
+        )
+        if not relation:
+            return
+
+        if user == "user":
+            username = self.karapace_requirer_user.fetch_relation_field(
+                relation_id=relation.id, field="username"
+            )
+            password = self.karapace_requirer_user.fetch_relation_field(
```

Comment:
**question** why not using the context? The idea of the pattern would be to centralize serialization/deserialization into a single place, and then use objects and attributes throughout

Reply:
It's the `app-charm` integration dummy, so no context here.

Reply:
ops, sorry! You are right! Not worth having the pattern there! Sorry